internal/alb/ls/listener.go

```diff
@@ -266,31 +266,59 @@ func (controller *defaultController) buildDefaultActions(ctx context.Context, op
 }
 
 // inferCertARNs retrieves a set of certificates from ACM that matches the ingress' hosts list
+// If multiple or none certificate were found for specific host, an error will be issued.
 func (controller *defaultController) inferCertARNs(ctx context.Context, ingress *extensions.Ingress) ([]string, error) {
 	var ingressHosts = uniqueHosts(ingress)
-	certArns := sets.NewString()
-
-	logger := albctx.GetLogger(ctx)
+	if len(ingressHosts) == 0 {
+		return nil, nil
+	}
 
-	certs, err := controller.cloud.ListCertificates([]string{acm.CertificateStatusIssued})
+	domainsByCertArn, err := controller.loadDomainsByCertArn(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	for _, c := range certs {
-		for _, h := range ingressHosts {
-			if domainMatchesHost(aws.StringValue(c.DomainName), h) {
-				logger.Infof("Domain name '%s', matches TLS host '%v', adding to Listener", aws.StringValue(c.DomainName), h)
-				certArns.Insert(aws.StringValue(c.CertificateArn))
-			} else {
-				logger.Debugf("Ignoring domain name '%s', doesn't match '%s'", aws.StringValue(c.DomainName), h)
+	certArns := sets.NewString()
+	for host := range ingressHosts {
+		certArnsForHost := sets.NewString()
+		for certArn, domains := range domainsByCertArn {
+			for _, domain := range domains {
+				if domainMatchesHost(domain, host) {
+					certArnsForHost.Insert(certArn)
+					break
+				}
 			}
 		}
+		if len(certArnsForHost) > 1 {
+			return nil, errors.Errorf("multiple certificate found for host: %s, certARNs: %v", host, certArnsForHost.List())
+		}
+		if len(certArnsForHost) == 0 {
+			return nil, errors.Errorf("none certificate found for host: %s", host)
+		}
+		certArns = certArns.Union(certArnsForHost)
 	}
-
 	return certArns.List(), nil
 }
 
+func (controller *defaultController) loadDomainsByCertArn(ctx context.Context) (map[string][]string, error) {
```

Comment:
The name is confusing, from the code, it is returning a map from `certArn` to list of `SAN`s. Should we name it as `loadCertArnToSANs`?